integrations/pgvector/src/haystack_integrations/document_stores/pgvector/document_store.py

```diff
@@ -0,0 +1,380 @@
+
+    def delete_table(self):
+        """
+        Deletes the table used to store Haystack documents.
+        """
+
+        delete_sql = SQL("DROP TABLE IF EXISTS {table_name}").format(table_name=Identifier(self.table_name))
+
+        self._execute_sql(delete_sql, error_msg=f"Could not delete table {self.table_name} in PgvectorDocumentStore")
+
+    def _handle_hnsw(self):
+        """
+        Internal method to handle the HNSW index creation.
+        It also sets the hnsw.ef_search parameter for queries if it is specified.
+        """
+
+        if self.hnsw_ef_search:
+            sql_set_hnsw_ef_search = SQL("SET hnsw.ef_search = {hnsw_ef_search}").format(
+                hnsw_ef_search=SQLLiteral(self.hnsw_ef_search)
+            )
+            self._execute_sql(sql_set_hnsw_ef_search, error_msg="Could not set hnsw.ef_search")
+
+        index_esists = bool(
+            self._execute_sql(
+                "SELECT 1 FROM pg_indexes WHERE tablename = %s AND indexname = %s",
+                (self.table_name, HNSW_INDEX_NAME),
+                "Could not check if HNSW index exists",
+            ).fetchone()
+        )
+
+        if index_esists and not self.hnsw_recreate_index_if_exists:
+            logger.warning(
+                "HNSW index already exists and won't be recreated. "
+                "If you want to recreate it, pass 'hnsw_recreate_index_if_exists=True' to the "
+                "Document Store constructor"
+            )
+            return
+
+        sql_drop_index = SQL("DROP INDEX IF EXISTS {index_name}").format(index_name=Identifier(HNSW_INDEX_NAME))
+        self._execute_sql(sql_drop_index, error_msg="Could not drop HNSW index")
+
+        self._create_hnsw_index()
+
+    def _create_hnsw_index(self):
+        """
+        Internal method to create the HNSW index.
+        """
+
+        pg_ops = SIMILARITY_FUNCTION_TO_POSTGRESQL_OPS[self.embedding_similarity_function]
+        actual_hnsw_index_creation_kwargs = {
+            key: value
+            for key, value in self.hnsw_index_creation_kwargs.items()
+            if key in HNSW_INDEX_CREATION_VALID_KWARGS
+        }
+
+        sql_create_index = SQL("CREATE INDEX {index_name} ON {table_name} USING hnsw (embedding {ops}) ").format(
+            index_name=Identifier(HNSW_INDEX_NAME), table_name=Identifier(self.table_name), ops=SQL(pg_ops)
+        )
+
+        if actual_hnsw_index_creation_kwargs:
+            actual_hnsw_index_creation_kwargs_str = ", ".join(
+                f"{key} = {value}" for key, value in actual_hnsw_index_creation_kwargs.items()
+            )
+            sql_add_creation_kwargs = SQL("WITH ({creation_kwargs_str})").format(
+                creation_kwargs_str=SQL(actual_hnsw_index_creation_kwargs_str)
+            )
+            sql_create_index += sql_add_creation_kwargs
+
+        self._execute_sql(sql_create_index, error_msg="Could not create HNSW index")
```

Comment:
Do we know why `Could not create HNSW index`? Any way we can add the actual error to these messages?

Reply:
In this case, we don't know...

this expression in `_execute_sql` should also show the actual error:
```python
try:
    result = cursor.execute(sql_query, params)
except Error as e:
    self._connection.rollback()
    raise DocumentStoreError(error_msg) from e
return result
```

Or did you mean something else?

Reply:
Ah no, shame on me, `from e` does what I want